### PR TITLE
Allow WidgetTab to use artwork tabs

### PR DESCRIPTION
**Pull request by StCyr**
This PR relates to https://github.com/LMMS/lmms/issues/2528

Br,

Cyrille

### CODE REVIEW

**Comment by tresf** [comment]
PR looks great.  We'll merge this once discussion over at #2528 is done.

**Comment by StCyr** [comment]
Damned, thought I had fixed this problem

**Comment by StCyr** [comment]
'Compiles without hicups here; Your build env is most probably more strict than mine.

There's most probably a problem there, bu I can't see it: The error is related with pointers and it's been a long time that I didn't write in C

**Comment by tresf** [comment]
The only Travis check failing is QT5 with the following error...  Qt4 on all other platforms is succeeding.

```
[ 99%] Building CXX object src/CMakeFiles/lmmsobjs.dir/gui/widgets/TextFloat.cpp.o
/home/travis/build/LMMS/lmms/src/gui/widgets/TabWidget.cpp: In member function ‘virtual void TabWidget::mousePressEvent(QMouseEvent*)’:
/home/travis/build/LMMS/lmms/src/gui/widgets/TabWidget.cpp:177:37: error: taking address of temporary [-fpermissive]
make[2]: *** [src/CMakeFiles/lmmsobjs.dir/gui/widgets/TabWidget.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [src/CMakeFiles/lmmsobjs.dir/all] Error 2
make: *** [all] Error 2
```

File: src/gui/widgets/TabWidget.cpp [hunk 1]
```diff
@@ -61,21 +66,31 @@ TabWidget::~TabWidget()
 
 
 
-void TabWidget::addTab( QWidget * _w, const QString & _name, int _idx )
+void TabWidget::addTab( QWidget * _w, const QString & _name, const QString & _tooltip, const char *activePixmap, const char *inactivePixmap, int _idx )
```
Comment (Umcaruje) [inline src/gui/widgets/TabWidget.cpp]: Function parameters are not prefixed with an underscore anymore, that's an old coding convention, so please change `_tooltip` to `tooltip` etc.

File: src/gui/widgets/TabWidget.cpp [hunk 2]
```diff
@@ -104,27 +119,73 @@ void TabWidget::setActiveTab( int _idx )
...
 
-
-
-void TabWidget::mousePressEvent( QMouseEvent * _me )
+// Return the index of the tab at position "pos"
+int TabWidget::findTabAtPos( const QPoint *pos )
 {
-	if( _me->y() > 1 && _me->y() < 13 )
+
+	if( pos->y() > 1 && pos->y() < m_tabbarHeight - 1 )
```
Comment (Umcaruje) [inline src/gui/widgets/TabWidget.cpp]: +1 removing magic numbers

File: src/gui/widgets/TabWidget.cpp [hunk 3]
```diff
@@ -104,27 +119,73 @@ void TabWidget::setActiveTab( int _idx )
...
 	{
-		int cx = ( ( m_caption == "" ) ? 4 : 14 ) +
-					fontMetrics().width( m_caption );
-		for( widgetStack::iterator it = m_widgets.begin();
-						it != m_widgets.end(); ++it )
+               int cx = ( ( m_caption == "" ) ? 4 : 14 ) + fontMetrics().width( m_caption );
```
Comment (Umcaruje) [inline src/gui/widgets/TabWidget.cpp]: Here you could replace 14 and 4 with predefined variables, to make changes easier down the road.

File: src/gui/widgets/TabWidget.cpp [hunk 4]
```diff
@@ -177,8 +238,8 @@ void TabWidget::paintEvent( QPaintEvent * _pe )
 		p.drawText( 5, 11, m_caption );
 	}
 
-	// Calculate the tabs' x (tabs are painted next to the caption)
-	int tab_x_offset = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
+       // Calculate the tabs' x (tabs are painted next to the caption)
+       int tab_x_offset = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
```
Comment (Umcaruje) [inline src/gui/widgets/TabWidget.cpp]: same comment as https://github.com/LMMS/lmms/pull/2599/files#r61166468

File: src/gui/widgets/TabWidget.cpp [hunk 5]
```diff
@@ -192,24 +253,66 @@ void TabWidget::paintEvent( QPaintEvent * _pe )
...
+
+		} else
 		{
-			p.setPen( QColor( 32, 48, 64 ) );
-			p.fillRect( tab_x_offset, 2, ( *it ).nwidth - 6, 10, cap_col );
+
+			// Highlight tab when active
+                	if( it.key() == m_activeTab )
+	                {
+				p.setPen( QColor( 32, 48, 64 ) );
```
Comment (Umcaruje) [inline src/gui/widgets/TabWidget.cpp]: This color should not be hardcoded, and you'd want to make this themeable via CSS.

File: src/gui/widgets/TabWidget.cpp [hunk 6]
```diff
@@ -177,8 +238,8 @@ void TabWidget::paintEvent( QPaintEvent * _pe )
 		p.drawText( 5, 11, m_caption );
 	}
 
-	// Calculate the tabs' x (tabs are painted next to the caption)
-	int tab_x_offset = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
+       // Calculate the tabs' x (tabs are painted next to the caption)
+       int tab_x_offset = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
 
 	QColor cap_col( 160, 160, 160 );
```
Comment (Umcaruje) [inline src/gui/widgets/TabWidget.cpp]: This color should also be themeable via css.

File: src/gui/widgets/TabWidget.cpp (hunk 1 above)
Comment (StCyr) [inline src/gui/widgets/TabWidget.cpp]: fixed in commit 8253235bebe3e3ba1eaf16b8e55d4bf29988083b

**Comment by StCyr** [comment]
Here's a mockup with a first attempt to flatten the TabWidget:

untitled

**Comment by Umcaruje** [comment]
This looks good.

Two remarks: the Filter background should be flattened too, so its all consistent now.

Could you please increase the Icon TabWidget height by 2 pixels? So the icons have 2 pixels of top and bottom padding. I think that would make it more pleasing for the eyes.

**Comment by Spekular** [comment]
Imo the TabWidget could be a bit darker so it stands out more from the
background.

**Comment by StCyr** [comment]
Hello @Umcaruje and @tresf 

Here's a screenshot as of my latest commit.

I removed the gradient from the GroupBox widget and added 2 pixels to the TabWidget's height.

untitled

**Comment by Umcaruje** [comment]
The spacing looks great now, but the highlighted part has a 1 px gap (marked red) now: 
screenshot from 2016-05-02 20 28 38

And I agree with @Spekular, the tab widget color should be darker and it should be themeable.

Here are some PR's that could help out with grasping the themeablity: #2547 #2442

**Comment by StCyr** [comment]
Hi @Umcaruje 

TabWidgets really are strange beasts.

Do you have any idea why, for example, the whole InstrumentTrackWindow turns red when I do the following small change?:

```c++
diff --git a/data/themes/default/style.css b/data/themes/default/style.css
index a66f779..bfc55a8 100644
--- a/data/themes/default/style.css
+++ b/data/themes/default/style.css
@@ -126,6 +126,11 @@ PianoRoll {
        qproperty-textShadow: rgb( 240, 240, 240 );
 }

+TabWidget {
+/*     background-color: #c9c9c9; */
+       background-color: rgb(255,0,0);
+}
+
diff --git a/src/gui/widgets/TabWidget.cpp b/src/gui/widgets/TabWidget.cpp
index 0dce731..331a2cc 100644
--- a/src/gui/widgets/TabWidget.cpp
+++ b/src/gui/widgets/TabWidget.cpp
@@ -209,16 +206,17 @@ void TabWidget::paintEvent( QPaintEvent * pe )
        setFont( pointSize<8>( font() ) );
        QPainter p( this );

-       QColor bg_color = QApplication::palette().color( QPalette::Active, QPalette::Background );
+//     QColor bg_color = QApplication::palette().color( QPalette::Active, QPalette::Background );
+       QBrush bg_color = p.background();

        p.fillRect( 0, 0, width() - 1, height() - 1, bg_color );

        bool big_tab_captions = ( m_caption == "" );

-       p.setPen( bg_color.darker( 150 ) );
+       p.setPen( bg_color.color().darker( 150 ) );
        p.drawRect( 0, 0, width() - 1, height() - 1 );

-       p.setPen( bg_color.light( 150 ) );
+       p.setPen( bg_color.color().light( 150 ) );
        p.drawLine( width() - 1, 0, width() - 1, height() - 1 );
        p.drawLine( 0, height() - 1, width() - 1, height() - 1 );
```

untitled

**Comment by StCyr** [comment]
Hello,

I wanted to let you know that I made some stuffs themeable in TabWidgets. Here's a mockup:

untitled

Things that are themeable are the following:

1. background color
2. Tabwidget's active tab's highlighiting color
3. TabWidget's title text color
4. Tabwidget's tab text color

Things that are not themeable at the moment:

1. various lines (background color + darker(150)) => I don't know if the various lines are still needed in a flatened design. What do you think?
2. Tabwidget's tabbar backgound color (background color + darker(150)) => Do you want it themeable?

Also, I don't find it very nice to have the TabWidget flatened while the GroupBox aren't. Maybe GroupBox should also be flatened?

Best regards,

Cyrille

**Comment by StCyr** [comment]
Ok, I've made a test without the lines.

I find it better when only the external borders are kept.

Here are 2 mockups: One without any lines, and the other with only the external borders:

untitled

untitled

**Comment by StCyr** [comment]
Here's a mockup with flatened GroupBox.

The background of the GroupBox's title bar is not themeable yet.

untitled

**Comment by IvanMaldonado** [comment]
If the GroupBox gets editable, along with the "GroupBoxes" of the other elements, it would help #2770 "Border styles"

**Comment by Umcaruje** [comment]
@StCyr Looks really cool, and much nicer. The highlighted tab should spread across the full height though.

**Comment by RebeccaDeField** [comment]
@StCyr Can we keep the icons white, so that the colors are not reversed when selected? I think it's cool, I just didn't see any other icons like that. Maybe we could make the selected background darker instead of lighter for a bit more contrast.

Great work!

**Comment by StCyr** [comment]
Hello @Umcaruje,

Is this better?

untitled
untitled

The problem is that the box under the artwork tab draws a line of the same color as the tab's background

@IvanMaldonado Do you mean that you want the borders to change color when selected?

@RebeccaDeField I will look how I could make all this themeable

**Comment by StCyr** [comment]
Hello,

I've made the TabWidget's title bar background themeable. Same for its borders.

Here's a mockup with "standards" colors and with @RebeccaDeField proposal of keeping the icons white:

untitled

I also investigated the possibility to have the icon colors themeable by changing the icons' color table when it is selected (eg: QImage::SetColor and http://www.qtcentre.org/threads/48211-How-to-change-QPixmap-color-from-black-to-red) but couldn't get any good results yet. It might also induce quite a high computing load.

Cheers,

Cyrille

**Comment by tresf** [comment]
> (eg: QImage::SetColor and http://www.qtcentre.org/threads/48211-How-to-change-QPixmap-color-from-black-to-red) but couldn't get any good results yet. It might also induce quite a high computing load.

If you do it every paint call, it can kill performance, but on mouse-over shouldn't be that bad.  Perhaps you generate the second color pixmap on construction and keep it cached and render in a flip/flop manner.

**Comment by StCyr** [comment]
Hello,

I've somewhat succeeded to make the artworks' colors themeable. Here are 2 mockups: The first one with "standard" colors and the second one with "extreme" colors to better show the "themeability":

untitled

untitled

One of the problem atm is that I can't use semi-transparent colors anymore, and thus the icons @RebeccaDeField made don't look as nice as before.

The following improvements might also be requested:

1. Fine tuning of the highlighting background's width, as the last tab's highlighting background overlaps a little on its right side
2. Make the GroubBox's background and borders also themeable
3. Fix the off-by-1pixel issue in the position of the "Triple Oscillator" pane, which, I think, is one pixel too tall
4. Other off-by-1pixel positioning issues that we couldn't see before because the TabWidget's background was the same color as the TabWidget's borders.

What do you think?

**Comment by StCyr** [comment]
For information, here's the part of the code responsible for changing the icons' color:

```cpp
// Create a mask out of the tab's artwork (for changing it's color later on)
QBitmap mask = QPixmap( embed::getIconPixmap( ( *it ).activePixmap ) ).createMaskFromColor( Qt::black, Qt::MaskOutColor );

// Select artwork's color and highlight active tab
if( it.key() == m_activeTab )
{
         p.fillRect( tab_x_offset, 0, ( *it ).nwidth, m_tabbarHeight - 1, tabSelected() );
         p.setPen( tabArtworkActive() );
} else
{
          p.setPen( tabArtworkInactive() );
}

// Draw colorized artwork
p.drawPixmap(tab_x_offset + ( ( *it ).nwidth - mask.width() ) / 2, 1, mask );
```

**Comment by tresf** [comment]
@StCyr I think it's safe to assume the artwork can live without custom masks (that inherently break alpha blending and are also bad on performance).  If we need a hover pixmap, I vote we simply add a second pixmap for now.  That way the icons are exactly as originally intended.

If we do want the icons to be colorized, it would probably be better to do so on the image itself, rather than a mask as masks can be problematic when dealing with alpha channels, however colorizing an image usually preserves alpha channel information.

http://stackoverflow.com/a/22152735/3196753 (Warning, comments seem to be inconclusive on Qt5)

**Comment by Umcaruje** [comment]
@StCyr I agree with @tresf, this is a bit of an overkill, and the lack of transparency is also a big problem, so I think we should just drop the idea. As far as the icon colors go, I like the version where they are just white.

**Comment by RebeccaDeField** [comment]
Thanks for all of the hard work @StCyr! I do agree that removing the transparency does not give the icons enough blending to look right.

Does the background change when you hover? That might be enough interaction for the user.

File: src/gui/widgets/TabWidget.cpp [hunk 7]
```diff
@@ -28,54 +28,73 @@
...
-}
+	// Create taller tabbar when it's to display artwork tabs
+	m_tabbarHeight = usePixmap ? GRAPHIC_TAB_HEIGHT : TEXT_TAB_HEIGHT;
+
+	m_tabheight = caption.isEmpty() ? m_tabbarHeight - 3 : m_tabbarHeight - 4;
 
+	setFont( pointSize<8>( font() ) );
 
+	setAutoFillBackground( true );
+        QColor bg_color = QApplication::palette().color( QPalette::Active,
```
Comment (tresf) [inline src/gui/widgets/TabWidget.cpp]: Some formatting here that should be fixed (tabs versus spaces).

File: src/gui/widgets/TabWidget.cpp [hunk 8]
```diff
@@ -28,54 +28,73 @@
...
-	m_widgets[_idx] = d;
-	_w->setFixedSize( width() - 4, height() - 14 );
-	_w->move( 2, 13 );
-	_w->hide();
 
+	// Tab's width when it is a text tab. This isn't correct for artwork tabs, but it's fixed later during the PaintEvent
+	int tab_width = fontMetrics().width( name ) + 10;
+
+        // Register new tab
+	widgetDesc d = { w, activePixmap, inactivePixmap, name, tooltip, tab_width } ;
```
Comment (tresf) [inline src/gui/widgets/TabWidget.cpp]: More tabs vs. spaces formatting here.

File: src/gui/widgets/TabWidget.cpp [hunk 9]
```diff
@@ -135,88 +200,104 @@ void TabWidget::resizeEvent( QResizeEvent * )
...
+		{	
+			// Fixes tab's width, because original size is only correct for text tabs
+			( *it ).nwidth = tab_width;
+
+			// Create a mask out of the tab's artwork (for changing it's color later on)
+			QBitmap mask = QPixmap( embed::getIconPixmap( ( *it ).activePixmap ) ).createMaskFromColor( Qt::black, Qt::MaskOutColor );
+
+			// Select artwork's color and highlight active tab
+			if( it.key() == m_activeTab )
+                        {
```
Comment (tresf) [inline src/gui/widgets/TabWidget.cpp]: More formatting.

File: src/gui/widgets/TabWidget.cpp [hunk 10]
```diff
@@ -135,88 +200,104 @@ void TabWidget::resizeEvent( QResizeEvent * )
...
+				p.fillRect( tab_x_offset, 0, ( *it ).nwidth, m_tabbarHeight - 1, tabSelected() );
+				p.setPen( tabArtworkActive() );
+			} else
```
Comment (tresf) [inline src/gui/widgets/TabWidget.cpp]: For consistency use:

```
}
else
{
```

Per https://github.com/LMMS/lmms/wiki/Coding-conventions

**Comment by tresf** [comment]
@StCyr if you have time to fix this up we can get it into the RC2 release.  There are some considerations in both code and implementation in the comments.  Once they're addressed, we'll merge.

**Comment by StCyr** [comment]
Hi @tresf 

I've adapted my code to respect you coding conventions; I believe it is ok now.

Do you see something else to do?

Please note that I'll be away for 2 weeks in Augustus.

File: data/locale/fr.ts [hunk 11]
```diff
@@ -3261,13 +3261,11 @@ Vous pouvez supprimer et déplacer les canaux d&apos;effet avec le menu contextu
...
+        <translation>FONCTIONS</translation>
     </message>
     <message>
         <source>FX</source>
-        <translatorcomment>&quot;EFFETS&quot; (previous translation) is too long for the limited available width. &quot;FX&quot; should be understable by all</translatorcomment>
-        <translation>FX</translation>
+        <translation>EFFETS</translation>
     </message>
     <message>
         <source>MIDI</source>
```
Comment (tresf) [inline data/locale/fr.ts]: @liushuyu will this commit cause problems with the translation service you're managing?

File: src/gui/widgets/TabWidget.cpp [hunk 12]
```diff
@@ -104,27 +123,74 @@ void TabWidget::setActiveTab( int _idx )
...
+
+	// Find index of tab that has been clicked
+	QPoint pos = me->pos();
+	int idx = findTabAtPos( &pos );
+
+	// When found, activate tab that has been clicked 
+	if ( idx != -1 )
+        {
+		setActiveTab( idx );
+                update();
```
Comment (tresf) [inline src/gui/widgets/TabWidget.cpp]: tab vs. spaces here

File: src/gui/widgets/TabWidget.cpp [hunk 13]
```diff
@@ -135,88 +201,106 @@ void TabWidget::resizeEvent( QResizeEvent * )
...
+			p.setPen( tabText() );
+			p.drawText( tab_x_offset + 3, m_tabheight + 1, ( *it ).name );
 		}
-		p.drawText( tab_x_offset + 3, m_tabheight, ( *it ).name );
-		p.setPen( cap_col );
+
+		// Next tab's horizontal position
 		tab_x_offset += ( *it ).nwidth;
-	}
+        }
```
Comment (tresf) [inline src/gui/widgets/TabWidget.cpp]: tabs vs. spaces here

File: src/tracks/InstrumentTrack.cpp [hunk 14]
```diff
@@ -1444,11 +1444,11 @@ InstrumentTrackWindow::InstrumentTrackWindow( InstrumentTrackView * _itv ) :
 	m_miscView = new InstrumentMiscView( m_track, m_tabWidget );
 
 
-	m_tabWidget->addTab( m_ssView, tr( "ENV/LFO" ), 1 );
-	m_tabWidget->addTab( instrumentFunctions, tr( "FUNC" ), 2 );
-	m_tabWidget->addTab( m_effectView, tr( "FX" ), 3 );
-	m_tabWidget->addTab( m_midiView, tr( "MIDI" ), 4 );
-	m_tabWidget->addTab( m_miscView, tr( "MISC" ), 5 );
+	m_tabWidget->addTab( m_ssView, tr( "ENV/LFO" ), tr( "ENV/LFO" ), "env_lfo_active", "env_lfo_inactive", 1 );
```
Comment (tresf) [inline src/tracks/InstrumentTrack.cpp]: should we be translating twice?  @liushuyu does this put unnecessary duplicates in our translation files or will this be consolidated?

Comment (tresf) [inline src/tracks/InstrumentTrack.cpp]: Ok, I see what is happening.  In this case, the tooltip is a new addition, but the name, is it still needed?  Can't we just make the name the tooltip when in pixmap tab mode, or does that break logic elsewhere?

Comment (jasp00) [inline src/tracks/InstrumentTrack.cpp]: Words in the tooltip should not be abbreviated.

Comment (StCyr) [inline src/tracks/InstrumentTrack.cpp]: Agree. LMMS would benefit having more descriptive tooltips.

What about?:

- ENV/LFO => "instrument's envelope and low-frequency oscillator parameters"
- FUNC =>"Chord and arpeggio settings"
- FX =>"Manage instrument's effects"
- MIDI => "Configure instrument's MIDI parameters"
- MISC => "Other instrument parameters"

File: src/gui/widgets/TabWidget.cpp [hunk 15]
```diff
@@ -135,88 +201,106 @@ void TabWidget::resizeEvent( QResizeEvent * )
...
 	{
-		p.setPen( QColor( 255, 255, 255 ) );
+		p.setFont( pointSize<8>( p.font() ) );
+		p.setPen( tabTitleText() );
 		p.drawText( 5, 11, m_caption );
 	}
 
-	// Calculate the tabs' x (tabs are painted next to the caption)
-	int tab_x_offset = m_caption.isEmpty() ? 4 : 14 + fontMetrics().width( m_caption );
+        // Calculate the tabs' x (tabs are painted next to the caption)
```
Comment (Umcaruje) [inline src/gui/widgets/TabWidget.cpp]: Tabs vs spaces issue here

File: src/gui/widgets/GroupBox.cpp [hunk 16]
```diff
@@ -99,35 +99,18 @@ void GroupBox::updatePixmap()
 	QColor bg_color = QApplication::palette().color( QPalette::Active,
```
Comment (Umcaruje) [inline src/gui/widgets/GroupBox.cpp]: Could you please make this background color themeable too? `p.background();` should do the trick again.

File: src/tracks/InstrumentTrack.cpp (hunk 14 above)
Comment (jasp00) [inline src/tracks/InstrumentTrack.cpp]: I suggest:
* ENV/LFO => "Envelope and low-frequency oscillator"
* FUNC =>"Chord and arpeggio"
* FX =>"Effects"
* MIDI => "MIDI"
* MISC => "Miscellaneous"

File: src/gui/widgets/GroupBox.cpp (hunk 16 above)
Comment (StCyr) [inline src/gui/widgets/GroupBox.cpp]: Then, I guess I'll also have to make the outer rectangle's color themeable also?

Or, would "p.background().color().darker( 150 )" sufficient?

**Comment by liushuyu** [comment]
Sorry for late to the party

>should we be translating twice? @liushuyu does this put unnecessary duplicates in our translation files or will this be consolidated?

No, since it is in the same context, I don't think it will be translated twice. Even this is the case, our translation hosting platform, Transifex, will handle that as well.

What I suggest here is to add translation comments or hints about the strings/context to inform the translators who are less familiar with coding. The comments (hints) can be added either using special code comments like `//:` (standard C++ style comments follow by a colon) or editing the translation template (in our case, that's our `en.ts` language file).

**Comment by Umcaruje** [comment]
@StCyr are you still available to work on this issue?

**Comment by StCyr** [comment]
Hi @Umcaruje 

Yes, I can still spend a few hours on this issue... But, I must say I'm completly lost in regard with what's still expected from me :-(

**Comment by Umcaruje** [comment]
Well, what I would first highly suggest is to actually rebase this pull request, as we have a new theme now. Plus we should work together to pick a look that fits on the new theme. @RebeccaDeField could probably help us on this regard.

Also you should adress this: https://github.com/LMMS/lmms/pull/2599/files#r75424006

And also, from what I saw, only one set of icons is used atm, so that could get cleaned up.

**Comment by Umcaruje** [comment]
Well, this PR has officially gone stale

**Comment by Umcaruje** [comment]
If I have time, I might revive this Pull Request

**Comment by Umcaruje** [comment]
Ok so I can't push to this branch, guess I'll need to make a new Pull request.

**Comment by Umcaruje** [comment]
Superseded by #3569

**Pull request closed without merging**